nix/nixos/svclib.nix

```diff
@@ -0,0 +1,328 @@
+  mkChairmanScript =
+    { chairman-config ? {
+        enable = true;
+      }
+    , cardano-cluster-config ? {
+        enable = true;
+        legacy-node-count = 4;
+        shelley-node-count = 3;
+        genesis-dir = "${../../configuration/b0109}";
+      }
+    ## NOTE:  we do the above defaulting, so the following works, FWIW
+    ##    nix-build -A scripts.mainnet.chairman
+    }:
+  let
+    injectServiceConfigs = {
+      config.services.chairman = chairman-config;
+      config.services.cardano-cluster = cardano-cluster-config;
+    };
+    pkgsModule = {
+      config._module.args.pkgs = mkDefault pkgs;
+    };
+    systemdCompat.options = {
+      systemd.services = mkOption {};
+      assertions = [];
+      users = mkOption {};
+    };
+    script = (modules.evalModules {
+      modules = [
+        ((import ../sources.nix).cardano-byron-proxy + "/nix/nixos")
+        ./cardano-cluster-service.nix
+        ./cardano-node-legacy-service.nix
+        ./cardano-node-service.nix
+        ./chairman-as-a-service.nix
+        injectServiceConfigs
+        pkgsModule
+        systemdCompat
+      ];
+    }).config.services.chairman.script;
+  in pkgs.writeScript "chairman" ''
+    #!${pkgs.runtimeShell}
+    set -euo pipefail
+    ${script} $@
+  '';
+
+  ## mkProxyFollowerTopology
+  ##   :: TopologyArgs -> FilePath (Topology Rewrite)
+  mkProxyFollowerTopology =
+    with builtins;
+    { hostAddr
+    , proxyPort
+    , nodePort
+    }:
+    let
+      topology = [{
+        nodeAddress = {
+          addr = hostAddr;
+          port = nodePort;
+        };
+        producers = [
+          { addr = hostAddr;
+            port = proxyPort;
+            valency = 1;
+          }
+        ];
+      }];
+    in toFile "proxy-follower-topology.yaml" (toJSON topology);
+
+  ## mkProxyScript
+  ##   :: ProxyOptions -> String ShellCmd
+  mkProxyScript =
+    { cfg, cardanoConfig, configurationKey, producers }:
+    ''
+      args=(
+        +RTS -T -RTS
+        --database-path ${cfg.stateDir}/db
+        --index-path ${cfg.stateDir}/index
+        --configuration-file ${cardanoConfig}/lib/configuration.yaml
+        --configuration-key ${configurationKey}
+        --topology ${cfg.topologyFile}
+        --logger-config ${cfg.logger.configFile}
+        --local-addr [${cfg.proxyHost}]:${toString cfg.proxyPort}
+        ${optionalString (cfg.pbftThreshold != null) "--pbft-threshold ${cfg.pbftThreshold}"}
+        ${optionalString (cfg.nodeId != null) "--node-id ${cfg.nodeId}"}
+        ${optionalString (cfg.listen != null) "--listen ${cfg.listen}"}
+        ${concatStringsSep " " (map (x: "--producer-addr ${x}") producers)}
+      )
+      echo "Starting cardano-byron-proxy:  ''${args[@]}"
+      exec ${cfg.package}/bin/cardano-byron-proxy "''${args[@]}"
+    '';
+  # ${optionalString (cfg.address != null) "--address ${cfg.address}"}
+  # --disable-peer-host-check
+in
+{ inherit
+  mkFullyConnectedLocalClusterTopology
```

Comment:
all these functions seem like they would be really beneficial to have all the time. Could we move this up a directory, call it nodeLib, and inherit nodeLib in lib.nix so it would be simple to use them all on the repl?

Reply:
@disassembler, pushed!